### DISCUSSION TITLE
Dependencies between tasks

### DISCUSSION THREAD

**Discussion by kirillsalykin** [maxcountryman/underway#18] [Ideas]
hi,

Huge thanks for the `underway`!
I've had implemented something similar to the `underway`, but in clojure.

We also needed a deps between jobs, for instance - before the email can be sent - pdf document should be generated; maybe this is something useful here as well?

it was pretty straightforward to implement, not complex DAG; based on table like this:
```
sql

CREATE TABLE jobs_deps (
  id SERIAL PRIMARY KEY,
  dependent_id INT NOT NULL REFERENCES jobs(id),
  dependency_id INT NOT NULL REFERENCES jobs(id),
  created_at timestamp with time zone NOT NULL DEFAULT NOW()
);

```

and it also changes how you pull next job - you want to ensure that there are no not completed dependend jobs.

probably it also requires to have something like `enqueue_with_deps(jobs: Vec<impl Job>).

WDYT?

**Comment by maxcountryman** [comment]
I've been working on an extension of the jobs abstraction that has exactly this in mind.

If you have time to take a look, I'd love your thoughts about the design in the [step functions branch](https://github.com/maxcountryman/underway/tree/step-functions).

**Reply by kirillsalykin** [reply]
I think steps is nice feature, but it is slightly different compared to deps I have in mind.

Imagine this situation: you have n jobs which process some data and then you a have job which generates report based on these n jobs.
with deps it is possible to cover this scenario, not sure if this is possible with steps.
tbh i've never experienced need in something like (there were plans, but didnt happen), so cannot say if this is something happens often.

Also, deps are connected but separate tasks, so each can be retried, afaik steps applied to the same task (I didnt found any changes in db design)

P.S.

the only thing that worries me - that tx becomes static, feels worrying... and as I understand this happens even if you dont need steps functionality.
I mean this:
```
        let step_tx: Transaction<'static, Postgres> = unsafe { mem::transmute_copy(&tx) };
```

which leads to removal of this :(
```
## 🦺 Safety

This crate uses `#![forbid(unsafe_code)]` to ensure everything is implemented in 100% safe Rust.
```

**Reply by kirillsalykin** [reply]
Another concern (if I understand everything correctly) - there is no way to commit tx within the step

**Reply by maxcountryman** [reply]
You can commit the tx.

The only concern around safety is if the tx is moved outside of the closure--it's possible but not easily done and it's far too valuable to have full atomicity so for now it's unsafe until we can find a way to propagate the correct lifetime through the trait object abstraction.

**Reply by maxcountryman** [reply]
Regarding "n jobs": this becomes n steps. Currently there's no way to run steps in parallel but something that could be considered in the future.

Stepwise functions are somewhat simpler than a true DAG and I'm partial to a simple interface here.

**Reply by kirillsalykin** [reply]
Seems like I am not fully understand how `steps` are working. Would appreciate a little description or diagram.

Thanks!

**Reply by maxcountryman** [reply]
https://github.com/maxcountryman/underway/blob/09b17ae4059022138d458b1fa04315099f0fae24/examples/step/src/main.rs#L42-L60

**Reply by kirillsalykin** [reply]
thank you

also see some nice changes, looking for a new release!

**Reply by kirillsalykin** [reply]
I am eager to try your changes already, volunteer to be your beta tester.
not sure I'll be using steps much tho...

**Reply by maxcountryman** [reply]
Thanks for that! I need to update the documentation and then should be able to cut a new 0.0.x release from the step functions branch. As you're testing, I'd love to read feedback you might have. 🙂

**Reply by maxcountryman** [reply]
@kirillsalykin I just published `0.0.5`. I also spent some time updating the documentation. In particular, if you have feedback about the job [module level docs](https://docs.rs/underway/latest/underway/job/index.html#), that would be helpful. Apart from that, if you run into any problems please do let me know. It's also worth mentioning that there's missing API still (unschedule, task cancellation, etc).

**Reply by kirillsalykin** [reply]
Will do, thanks

**Reply by kirillsalykin** [reply]
> [Tasks](https://docs.rs/underway/latest/underway/index.html#tasks) represent a well-structure unit of work.

I think there is a small typo, `well structure**d**` should it be?

**Reply by kirillsalykin** [reply]
the docs are looking awesome, clean and clear!

Also, I must admit `step`'s are pretty cool.
